runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,431 @@
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Commit the transaction.
+     * For a transaction that only has write operations, this method will start and end
+     * a corfu transaction keeping the "critical" section small while batching up all updates.
+     * For a transaction that has reads, this will end the transaction.
+     * The commit returns successfully if the write transaction was committed.
+     * Otherwise this throws a TransactionAbortedException with a cause.
+     * The cause and the caller's intent of the transaction can determine if this aborted
+     * Transaction can be retried.
+     * @return - address at which the commit of this transaction occurred.
+     */
+    public long commit() {
+        operations.forEach(Runnable::run);
+        long commitAddress = Address.NON_ADDRESS;
+        try {
+            commitAddress = this.objectsView.TXEnd();
+        } catch (Exception ex) {
+            tablesInTxn.forEach(t -> t.getMetrics().numTxnAborts.incrementAndGet());
+            tablesInTxn.clear();
+            throw ex;
+        }
+        long timeElapsed = System.nanoTime() - txnStartTime;
+        switch (txnType) {
+            case READ_ONLY:
+                tablesInTxn.forEach(t -> t.getMetrics().readOnlyTxnTimes.set(timeElapsed));
+                break;
+            case WRITE_ONLY:
+                tablesInTxn.forEach(t -> t.getMetrics().writeOnlyTxnTimes.set(timeElapsed));
+                break;
+            case READ_WRITE:
+                tablesInTxn.forEach(t -> t.getMetrics().readWriteTxnTimes.set(timeElapsed));
+                break;
+            default:
+                log.error("UNKNOWN TxnType!!");
+                break;
+        }
+        tablesInTxn.clear();
+        operations.clear();
+        return commitAddress;
+    }
+
+    /**
+     * Explicitly abort a transaction in case of an external failure
+     */
+    public void txAbort() {
+        operations.clear();
+        tablesInTxn.clear();
+        if (TransactionalContext.isInTransaction()) {
+            this.objectsView.TXAbort();
+            tablesInTxn.forEach(t -> t.getMetrics().numTxnAborts.incrementAndGet());
+        }
+    }
+
+    /**
+     * Cleanup the transaction resources.
+     * If invoked on transaction with just queries, record the time taken.
+     */
+    @Override
+    public void close() {
+        if (TransactionalContext.isInTransaction()) {
+            if (txnType == READ_ONLY) {
+                long timeElapsed = System.nanoTime() - txnStartTime;
+                tablesInTxn.forEach(t -> t.getMetrics().readOnlyTxnTimes.set(timeElapsed));
+            } else {
+                log.warn("close()ing a transaction with mutations without calling commit()!");
+            }
+            this.objectsView.TXAbort();
```

Comment:
I think if you want to support the case that user can end read-only transaction without calling commit, then it's better to use `this.objectsView.TxEnd()` instead of `this.objectsView.TXAbort()`, because maybe Corfu internally bookkeeps some metrics or having logs, if using TXAbort, that will confuse the corfu logging and metrics. I personally think we should not support this, because it seems error prone: imagining the txn was intially read-only and user didn't write commit(), later they modifiied the transaction and added some write logic, but forget to add commit(), then since we only log a warning here, it will silently fail. I think we should encourage standarizing the transaction commit pattern to avoid confusion and possible mistakes.

Reply:
Yes, but Autcoclosable in query was an ask. will add TxEnd() but that's about the best we can do I guess.

Reply:
I think we don't need to satisfy every ask from the applications, for this one, they just want a convenience, but that convenience is breaking our commit pattern and can lead to more problems, I feel it's not worth to do that...